tempesta_fw/ss_skb.c

```diff
@@ -92,72 +93,93 @@ ss_skb_alloc_pages(size_t len)
+ *   kernel's allocators, then that page is owned by the allocator. That
+ *   page's reference count will not get down to 1 until the page is released
+ *   by the allocator.
+ * - Tempesta's allocator has no notion of "owning" a page that is used
+ *   for allocation of fragments. These fragments are used for the SKB
+ *   structure itself, as well as for skb->head. The page used for these
+ *   parts of an SKB may propagate to a paged fragment when a part of the
+ *   linear data is mapped to a fragment. The algorithm skips those pages to
+ *   avoid an unwanted memory corruption.
+ * - map[] array is used to speed up the search. Once fragments that belong
+ *   in the same page are checked, there's no need to check them again in the
+ *   next iteration.
  */
-static skb_frag_t *
-__check_frag_room(struct sk_buff *skb, skb_frag_t *frag, int len)
+static bool
+__lookup_pgfrag_room(const struct sk_buff *skb, int len, skb_frag_t *f_out)
 {
-	int i, sz, sz2, refcnt;
-	struct page *pg = skb_frag_page(frag);
-	skb_frag_t *frag2, *ret = frag;
+	int i, k, owners;
+	char map[MAX_SKB_FRAGS];
+	const skb_frag_t *f_base, *f_this;
+	unsigned int p_size, h_room, t_room;
+	struct page *p_base, *p_skb_head = virt_to_head_page(skb->head);
 
-	if ((refcnt = page_count(pg)) == 1)
-		return frag; /* no other users */
+	memset(map, 0, sizeof(map));
 
-	sz = PAGE_SIZE - ss_skb_frag_len(frag);
-	for (i = skb_shinfo(skb)->nr_frags - 1; i >= 0 ; --i) {
-		frag2 = &skb_shinfo(skb)->frags[i];
-		if (frag2 == frag || pg != skb_frag_page(frag2))
+	for (i = 0; i < skb_shinfo(skb)->nr_frags; i++) {
+		if (map[i])
 			continue;
-		sz2 = PAGE_SIZE - ss_skb_frag_len(frag2);
-		if (sz2 < len)
-			return NULL;
-		if (sz2 < sz) {
-			sz = sz2;
-			ret = frag2;
-		}
-		/* Return localy referenced pages only. */
-		if (--refcnt == 1)
-			return ret;
-	}
 
-	/* The page is used outside of this SKB. */
-	return NULL;
-}
+		f_base = &skb_shinfo(skb)->frags[i];
+		p_base = compound_head(skb_frag_page(f_base));
+		if (p_base == p_skb_head)
+			continue;
 
-/**
- * Look up a page fragment that has room for @len bytes.
- */
-static skb_frag_t *
-__lookup_pgfrag_room(struct sk_buff *skb, int len)
-{
-	int i;
+		owners = page_count(p_base);
+		p_size = h_room = t_room = PAGE_SIZE << compound_order(p_base);
 
-	/*
-	 * Iterate in reverse order to use likely moving fragments.
-	 * Thus we find free room more frequently and skb fragments
-	 * utilize memory limits better.
-	 */
-	for (i = skb_shinfo(skb)->nr_frags - 1; i >= 0; --i) {
-		skb_frag_t *frag = &skb_shinfo(skb)->frags[i];
-		if ((int)PAGE_SIZE - ss_skb_frag_len(frag) < len)
-			continue;
-		frag = __check_frag_room(skb, frag, len);
-		if (frag)
-			return frag;
+		for (k = i; k < skb_shinfo(skb)->nr_frags; k++) {
+			if (map[k])
+				continue;
+
+			f_this = &skb_shinfo(skb)->frags[k];
+			if (compound_head(skb_frag_page(f_this)) != p_base)
+				continue;
+
+			if ((map[k] = 1, map[i]) > 1)
+				continue;
+
```

Comment:
The first iteration of the loop always goes here and it doesn't need to pass the 3 conditions, so it's better just to initialize `map[i]` before the loop.